### DISCUSSION TITLE
3D programming related questions

### DISCUSSION THREAD

**Discussion by Mazaaap** [axmolengine/axmol#1533] [Q&A]
Axmol newbie here. For OpenGL there seems to be plenty of examples available, but Axmol is giving me some headaches. I have been planning a game that has a 3D world made from one big object (c3b) and then adding additional objects for player movement, etc.

I would appreciate feedback on the following topics:

1. Enabling transparency for a texture alpha channel: Would looping through all nodes and checking if their names contain the string "X" to then call getMaterial and set setTransparent(true) + getStateBlock().setCullFaceSide(CullFaceSide::NONE) be the correct and most efficient approach?

2. Reflections for the combination of everything visible on the screen: There is an example under the cpp-tests directory (47: Node: MeshRenderer -> CubeMap & Skybox Test), but how would I make it more complex to handle multiple objects on the screen?

3. Shadow mapping: How would I generate a depth buffer (framebuffer?) from the point of view of a light source (sun)?

**Comment by solan-solan** [comment]
If it is still relevant:

1. Yes I think this approach is good
2. To my understand you should extend axmol fragment shader which responsible for 3d object rendering where you will pass skybox texture cube. Then skybox color should be added to the result color of this shader. 
3. Check please this example. It was not updated to the final axmol version, but it has sample of
    - creating separate depth camera
    - creating render texture with depth buffer binding
    - applying this buffer to render shadow whith custom object shaders
   It is just base implementation which includes necessary things. 

https://github.com/solan-solan/HeightMap/tree/smooth_lod_passing

**Comment by Mazaaap** [comment]
Thanks for the reply! Any help would be appreciated to convert the shaders Axmol 2.1 compatible since they crash at the start of the program:
```
Assertion failed!
Program: ...
File: \axmol\core\renderer\backend\opengl\ProgramGL.cpp
Line: 392
Expression: uniform.type == GL_SAMPLER_2D || uniform.type == GL_SAMPLER_CUBE```

**Comment by solan-solan** [comment]
Looks like you try building glsl 1.0 shader for axmol glsl 3.0 configuration?

**Comment by solan-solan** [comment]
It was updated to the final axmol version, shader's migrated to glsl 3.0
